forum/qa-plugin/user-activity-log/user-activity-log-search.php

```diff
@@ -0,0 +1,177 @@
+        $qa_content['title']= qa_lang_html('user-activity-log/searchResults').$this->searchArr['request'];
+        
+        $qa_content['custom'] = $this->generateResultsTable();
+        $qa_content['navigation']['sub']=qa_admin_sub_navigation();
+        return $qa_content;
+    }
+
+    private function dbSearch()
+    {
+        $baseQuery = 'SELECT `datetime`, `ipaddress`, `handle`, `event` FROM qa_eventlog ';
+
+        if($this->searchArr['condition'] === 'username'){
+            $finalQuery = $baseQuery.'WHERE  handle = $';
+        }else if($this->searchArr['condition'] === 'type'){
+            $finalQuery = $baseQuery.'WHERE event = $';
+        }else if($this->searchArr['condition'] === 'ip'){
+            if(qa_get_logged_in_level() > QA_USER_LEVEL_EDITOR){
+                $finalQuery = $baseQuery.'WHERE ipaddress = $';
+            }   
+        }
+
+        if(!empty($this->searchArr['date'])){
+            $dateArr = explode('-', $this->searchArr['date']);
+
+            if(isset($dateArr[0]) && is_numeric($dateArr[0])){
+                $date = $dateArr[0];
+                if(isset($dateArr[1]) && is_numeric($dateArr[1])){
+                    $date = $date.'-'.$dateArr[1];
+                    if(isset($dateArr[2])){
+                        if(strlen($dateArr[2]) > 2){
+                            [$day, $hours] = explode(' ', $dateArr[2]);
+                            $dateArr[2] = $day;
+                            $dateArr[3] = $hours;
+                        }
+                        
+                        $date = $date.'-'.$dateArr[2];
+                    }
+                }
+                $finalQuery = $finalQuery.'AND `datetime` LIKE $';
+            }
+
+        }
+        
+        
+        if(!is_numeric($this->searchArr['resultsCount'])){
+            $this->dbResult = [];
+            return 0;
+        }
+        
+        $finalQuery = $finalQuery.' ORDER BY `datetime` DESC LIMIT #';
+        if(strpos($finalQuery, 'LIKE')){
+            $result = qa_db_query_sub($finalQuery, $this->searchArr['request'], $this->searchArr['date']."%", $this->searchArr['resultsCount']);
+        }else{
+            $result = qa_db_query_sub($finalQuery, $this->searchArr['request'], $this->searchArr['resultsCount']);
+        }
+       
+        if(mysqli_num_rows($result) <= $this->searchArr['resultsCount']){
+            $fullResultsArray = [];
+            while($row = mysqli_fetch_array($result, MYSQLI_ASSOC)){
+                $fullResultsArray[] = $row;
+            }
+            $this->dbResult = $fullResultsArray;
+            return 1;
+        }else{
+            $this->dbResult = $result;
+            return 0;
+        }
+
+    }
+
+    private function generateResultsTable()
+    {
+        $unfiltered = $this->dbResult;
+        if(qa_get_logged_in_level() < QA_USER_LEVEL_ADMIN){
+           $results = array_filter($unfiltered, function($field){
+                switch($field['event']){
```

Comment:
Chyba nie do końca rozumiem zamysł tego switcha. On ma po prostu wyfiltrować, aby wyświetlone zostały tylko wyniki z wymienionymi tu eventami?

Reply:
Zamysł tego switcha był taki

Jako jedno z uprawnień mamy "kto głosował bądź oflagował wiadomość", u nas widzą to jedynie administratorzy i wyżej. Ja zinterpretowałem 'wiadomości' jako 'pytania, komentarze, odpowiedzi'.

Ten switch usuwa nam z wyników informacje o flagach i głosach jeśli sprawdzający nie jest zalogowany jako admin lub wyżej.

Reply:
Ok, w takim razie ja osobiście wrzuciłbym to w jakieś `in_array()` i prostym warunkiem sprawdził, wydaje mi się to czytelniejsze niż seria case, które sprowadzają się i tak do jednej linii akcji. No i powstaje w takiej sytuacji jeszcze jeden mały problem: filtrowanie wyników z eventów odbywa się już po zapytaniu do bazy. A więc gdy redaktor poprosi o 20 wpisów, z czego 5 to będą wpisy dotyczące głosów, to tak naprawdę zobaczy ich finalnie 15?